Readme.md
# koa-dir-router[![NPM version](https://img.shields.io/npm/v/koa-dir-router.svg?style=flat)](https://npmjs.com/package/koa-dir-router)

[中文文档](http://koadirrouter.yamjs.cn/Readme_cn)

A middleware of KOA supports the access form with folder directory as the URL, and does not need to restart koa service. It can update the file code of the folder inside the folder and modify it. It is convenient to modify the code and go online without feeling when developing the interface.

You can use other middleware or define page mix. Koa dir router only works when `ctx.response.status===404`

## Installation

```
$ npm install koa-dir-router
```

## Options

- `dir` The absolute path to the directory where the code is stored, must `[String]`
- `prefixUrl` the prefix of the request path, which is used to match the local file. The default value is'/' `[String]`
- `checkTimes` Time to detect file changes, in MS; default is `1000` `[Number]`
- `errorLog` Code method caught when the file code under the file directory executes an exception; the received value is `[Function]({path,des,error})`
- `page404` When the file code in the file directory does not exist, the callback function receives a value of `[function](ctx)`
- `debug` Whether to display debugging information; the default value is `true` , the received value is `[Boolean]`
  -The parameter `acceptmethods` `[string]` supports setting the accepted method. The default value is '\*', and the specification is' get, post '(1.1.6 +)

-The parameter `httpmethod` `[array]` supports extended detection. The default method supported is`[get, post, put, delete]`(1.1.6 +)

> In version 1.0.7, the `baseurl` parameter name is abolished and replaced with `prefixurl`;

## Example

Start in the same level directory of the startup program

directory structure

```
·
├──index.js
├──controller
   ├──mis
      ├──type.js
```

```js
// ./controller/mis/type.js
module.exports = function (ctx) {
  ctx.body = `show-ok`
}
// ./index.js
const dirRouter = require('koa-dir-router')
const Koa = require('koa')
const path = require('path')
var app = new Koa()

app.use(
  dirRouter({
    dir: path.join(__dirname, './controller'), // Incoming directory structure to access
  })
)
app.listen(3000)
```

When accessing http://localhost:3000/mis/type, it will correspond to the file code of `./controller/MIS/type.JS` under the corresponding file directory

yields:

```js
$ GET /mis/type

show-ok
```

![showOK.png](https://static.bestsloth.top/showOk.png)

If you need a base address `baseUrl`

```js
// ./index.js
const dirRouter = require('koa-dir-router')
const Koa = require('koa')
const path = require('path')
var app = new Koa()

app.use(
  dirRouter({
    dir: path.join(__dirname, './controller'), // Incoming directory structure to access
    baseUrl: '/mis', // base address
  })
)
app.listen(3000)
```

When accessing http://localhost:3000/mis/mis/type, it will correspond to the file code of `./controller/MIS/type.JS` under the corresponding file directory
yields:

```js
$ GET /mis/mis/type

show-ok
```

![showOK2.png](https://static.bestsloth.top/showOk2.png)
，
**When something goes wrong with the code?**

If there is something wrong with the code during development, 'koa dir router' will have a friendly prompt

![error.png](https://static.bestsloth.top/error.png)

If it is online code, you can use 'errorlog' to get it.

## ** set the specified transmission mode (1.1.6 +)**

Set the parameter `acceptmethods` to specify the program to work in a specific transmission mode

## ** specify different execution method bodies for different request methods (1.1.6 +)**

You can use it in the code file`ctx.dirRouter.MethodsName`Method to specify the code body that is not used in different ways. The value range of methodsname is the file code `httpmethod`, provided that the mode is allowed to enter the main program with `acceptmethods`;

In this way, developers can quickly process the program in different ways

```js
module.exports = async function (ctx) {
  await ctx.dirRouter.get((c) => {
    console.log('run Get')
  })
  await ctx.dirRouter.post((c) => {
    console.log('run POST')
  })
}
```

# License

MIT


History.md
# 1.1.6 / 2020-07-22

- 增加`acceptMethods` `[String]`参数，支持设定接受的方法 默认是 '\*' ，规范是 'get,post'
- 增加`httpMethod` `[Array]`参数，支持扩展检测支持的方法 默认是`[get,post,put,delete]`
- 增加`ctx.dirRouter`增加方法捕获的事件【'暂仅支持 get,post,put,delete‘允许进行扩展 httpMethod】允许在方法内针对不同的请求方法进行分别处理；前提；该方法在`httpMethod`内包含，所择会提示错误

  ```js
  module.exports = async function (ctx) {
    await ctx.dirRouter.get((c) => {
      console.log('执行了Get')
    })
    await ctx.dirRouter.post((c) => {
      console.log('执行了POST')
    })
  }
  ```

# 1.1.3 / 2020-04-16

- 修复 bug，优化一场捕获

# 1.0.7 / 2020-03-3

- add prefixUrl,delete baseUrl
- remove `fs-extra` module

# 1.0.6 / 2020-03-1

- add debug, page404

# 1.0.0 / 2020-02-28

- add koa-dir-router


index.js
/*
 * @Author: xuxueliang
 * @Date: 2020-02-28 14:40:00
 * @LastEditors: xuxueliang
 * @LastEditTime: 2020-07-22 12:51:40
 */
const path = require('path')
const fs = require('fs')
const clearModule = require('clear-module')
const version = require('./package.json').version

const generateRunMethodsCall = (ctx) => {
  return async (cb) => {
    typeof cb === 'function' && cb(ctx)
  }
}

const httpMethods = ['get', 'post', 'put', 'delete']
/**
 * @method dirRouter
 * @param {String} dir 程序目录
 * @param {String} baseUrl 基准目录
 * @param {String} prefixUrl 基准目录
 * @param {Number} checkTimes 检查间隔ms
 * @param {Boolean} debug 是否显示调试信息
 * @param {Functin} errorLog 异常捕获记录
 * @param {String} acceptMethods 接受的方法 默认是* 所有方法都接受
 * @param {Array} httpMethod 允许的http传输类型
 * @param {Functin} page404 404页面
 * @param {Functin} context 函数调用的上下问对象 默认时global
 * @return {null} 
 * */
module.exports = ({
  dir = null, // 程序目录
  baseUrl = '',// 基准目录
  prefixUrl = '/',// 基准目录
  checkTimes = 1000,// 检查间隔
  debug = true,//是否显示调试信息
  errorLog = () => { },
  context = global,
  httpMethod = [],
  acceptMethods = '*',
  page404 = () => { }
} = {}) => {
  return async (ctx, next) => {
    await next()
    if (typeof acceptMethods === 'string') {
      acceptMethods = acceptMethods.toLowerCase()
    } else {
      acceptMethods = '*'
    }
    const methods = ctx.request.method.toLowerCase()

    if (acceptMethods.indexOf('*') < 0 && acceptMethods.indexOf(methods) < 0) {
      debug && console.log(`路由设置只接受：【${ acceptMethods }】，不接受【${ methods }】`)
      return
    }
    if (Array.isArray(httpMethod)) {
      httpMethods.push(...httpMethod)
    }
    const dirRouterObj = {
      debug, errorLog, dir,
    }
    ctx.dirRouter = new Proxy(dirRouterObj, {
      get: function (target, key, receiver) {
        if (key in target) {
          return Reflect.get(target, key, receiver);
        } else if (typeof key === 'string' && httpMethods.indexOf(key) > -1) {
          if ((acceptMethods.indexOf('*') > -1 || acceptMethods.indexOf(key) > -1) && methods === key) {
            return target[key] = generateRunMethodsCall(ctx)
          } else {
            return target[key] = () => { }
          }
        }
        return undefined
      }
    })
    if (ctx.response.status === 404 && dir) {
      let filePath = path.join(dir, getFilePath(ctx.request.url, baseUrl || prefixUrl))
      let requirepath = path.relative(__dirname, filePath)
      try {
        let data = require(requirepath)

        let mData = null
        if (!data._ckTime || Date.now() - data._ckTime >= checkTimes) {
          data._ckTime = Date.now()
          let _filePath = filePath + '.js'
          if (!fs.existsSync(_filePath)) {
            _filePath = filePath + '/index.js'
          }
          mData = await fs.statSync(_filePath)
          const mtime = mData.mtime.toString()
          if (!data._mtime) {
            data._mtime = mtime
          }
          if (data._mtime !== mtime) {
            clearModule(requirepath)
            data = require(requirepath)
            data._ckTime = Date.now()
            data._mtime = mtime
          }
        }
        try {
          if (typeof data === 'function') {
            await data.call(context, ctx)
          } else {

            handleError(new Error(filePath + '不是一个函数'), filePath, ctx)
          }
        } catch (e) {
          handleError(e, filePath, ctx)
        }
        data = null
        mData = null
      } catch (e) {

        if (e.toString().indexOf('no such file')) {
          ctx.body += `${ ctx.request.url }  链接不存 \r`
        } else {
          handleError(e, filePath, ctx)
        }
        if (ctx.app.env === 'development') {
          console.log(e)
        }
        if (typeof page404 === 'function') {
          page404(ctx) // 404页面需要自定义
        }
      }
    }
  }
}
function handleError (e, filePath, ctx) {
  console.log('文件 【' + filePath + '】 执行有问题')
  typeof ctx.dirRouter.errorLog === 'function' && ctx.dirRouter.errorLog({
    path: filePath,
    des: '文件 【' + filePath + '】 执行有问题',
    error: e
  })
  ctx.type = 'text/html;charset=utf-8'
  if (ctx.dirRouter.debug) {
    ctx.body = `<div><h3>【koa-dir-router】捕获的异常信息 </h3>`
    ctx.body += "<hr>"
    ctx.body += "错误名称: " + e.name + '<br>'
    ctx.body += "<pre>错误信息: " + e.stack.replace(ctx.dirRouter.dir, '【koa-dir-router的工作目录下的】') + '</pre>'
    ctx.body += "<hr>"
    ctx.body += "<em style='font-size:12px;color:#999'>要想屏蔽该报错信息，需要设置【koa-dir-router】参数[debug]为[false] </em>"
  } else {
    ctx.body = `${ ctx.request.url }  访问出错\r`
    ctx.body += "<hr>"
  }
  ctx.body += `<h5 style="font-size:16px;text-align:left;color:#999"><a target='_blank' href='https://www.npmjs.com/package/koa-dir-router'>koa-dir-router@${ version } 提供路由服务</a></h5></div>`
}
function getFilePath (url, prefixUrl) {
  prefixUrl = prefixUrl === '/' ? '' : prefixUrl
  return url.split('?')[0].substr(prefixUrl.length)
}
